engine/kv_engine_cleaner.cpp

```diff
@@ -235,366 +237,432 @@ void KVEngine::cleanNoHashIndexedSkiplist(
-    size_t slot_num = 0;
+  {  // purge and free pending old dl records
+    while (!pending_clean_records.pending_purge_dls.empty()) {
+      auto& pending_dls = pending_clean_records.pending_purge_dls.front();
+      if (pending_dls.release_time <
+          version_controller_.LocalOldestSnapshotTS()) {
+        purgeAndFreeDLRecords(pending_dls.records);
+        pending_clean_records.pending_purge_dls.pop_front();
+      } else {
+        break;
+      }
+    }
+  }
+
+  {  // Destroy skiplist
+    while (!pending_clean_records.outdated_skip_lists.empty()) {
+      auto& ts_skiplist = pending_clean_records.outdated_skip_lists.front();
+      if (ts_skiplist.first < version_controller_.LocalOldestSnapshotTS()) {
+        ts_skiplist.second->DestroyAll();
+        removeSkiplist(ts_skiplist.second->ID());
+        pending_clean_records.outdated_skip_lists.pop_front();
+      } else {
+        break;
+      }
+    }
+  }
+
+  {  // Destroy list
+    while (!pending_clean_records.outdated_lists.empty()) {
+      auto& ts_list = pending_clean_records.outdated_lists.front();
+      if (ts_list.first < version_controller_.LocalOldestSnapshotTS()) {
+        listDestroy(ts_list.second.release());
+        pending_clean_records.outdated_lists.pop_front();
+      } else {
+        break;
+      }
+    }
+  }
+
+  {  // Destroy hash
+    while (!pending_clean_records.outdated_hash_lists.empty()) {
+      auto& ts_hlist = pending_clean_records.outdated_hash_lists.front();
+      if (ts_hlist.first < version_controller_.LocalOldestSnapshotTS()) {
+        hashListDestroy(ts_hlist.second.release());
+        pending_clean_records.outdated_hash_lists.pop_front();
+      } else {
+        break;
+      }
+    }
+  }
+}
+
+size_t KVEngine::FetchCachedOutdatedVersion(
+    PendingPrugeFreeRecords& pending_clean_records,
+    std::vector<StringRecord*>& purge_string_records,
+    std::vector<DLRecord*>& purge_dl_records) {
+  size_t outdated_records_num = 0;
+  size_t i = round_robin_id_.fetch_add(1) % configs_.max_access_threads;
+  auto& tc = cleaner_thread_cache_[i];
+  std::deque<CleanerThreadCache::OutdatedRecord<StringRecord>>
+      outdated_string_records;
+  std::deque<CleanerThreadCache::OutdatedRecord<DLRecord>> outdated_dl_records;
+  if (tc.outdated_dl_records.size() > kMaxCachedOldRecords ||
+      tc.outdated_string_records.size() > kMaxCachedOldRecords) {
+    std::lock_guard<SpinMutex> lg(tc.mtx);
+    if (tc.outdated_string_records.size() > kMaxCachedOldRecords) {
+      outdated_string_records.swap(tc.outdated_string_records);
+    }
+    if (tc.outdated_dl_records.size() > kMaxCachedOldRecords) {
+      outdated_dl_records.swap(tc.outdated_dl_records);
+    }
+  }
+  if (outdated_string_records.size() > 0) {
+    outdated_records_num += outdated_string_records.size();
+    std::vector<StringRecord*> to_free_strings;
     version_controller_.UpdateLocalOldestSnapshot();
+    auto release_time = version_controller_.LocalOldestSnapshotTS();
```

Comment:
release_time is now()

Reply:
These code is from last PR(`cfe7aa0e68a9e40945831d8256d0a5b429969480`) , I just wrapper these into function.  Our project has many `release_time` name , please see `PendingPurgeDLRecords` and so on.  So I tend to keep here and don't change it.